Feature/23 feature request pin file loading (#24)
ynput/ayon-usd-resolver: Docs/Ayon_Docs/Dev_Docs.md

Author: moved to https://github.com/ynput/ayon-usd-resolver/pull/25#issuecomment-2343281677

# Dev Docs

### Into

Welcome to the Devloper Docs. this document will be split into 2 portions.

1: documentation for **AYON** developers that want to use the resolver but dont
want to work on the source code. it will outile all the things the resolver
offers and how you can controle them.\
2: the **Resolver Developer** part. this portion will talk about what we use for

> the resolver how we use it and

Author: I rewrote this sentence.
Because I'm not entirely sure what I wanted to say there.

context via `GetConnectedContext`\
this is also the recommended way to get access to the dropCache function for the
resolver

```py
explicit_resolver = AyonUsdResolver.Resolver()
explicit_resolver_context = explicit_resolver.GetConnectedContext()
```

`from usdAssetResolver import AyonUsdResolver`python bindings for the
AyonUsdResolver.

`Ar.SetPreferredResolver("AyonUsdResolver")` this command allows you to set a
preferred Resolver. if you use the AyonUsd addon you wont need to set this line
as we register the resolver as the main resolver. but if you have a different
resolver set as Main you might need to use this command.

`resolver = Ar.GetResolver()` there are multiple ways to get access to an
resolver instance but using the "Default" Usd option is usually the best idea as
it will return the resolver for the current context.

`context = AyonUsdResolver.ResolverContext()` accessing an ResolverContext
instance is useful if you want to manipulate the Global resolver cache.

### Pinning

the AyonUsdResolver has a feature called Pinning Support.\
in short pinning support allows you to load a pinning File and disconnect the

> AyonCppApi.\

Author: I added a short line of text to explain that a memory cache bypassed the resolver.

> AyonCppApi.\

Author: i belive this is done so im gonna close it. 
PS: im moving the issues over to the docs branch just so you know if you want to open it again.

why would you do that ? When running a resolver on a farm with many many
systems(Render Nodes) you might not want them all to call the Ayon server to
avoid impacting the server performance.

**How to use it ?**\
There are 3 Env variables that control the Pinning file support

> `ENABLE_STATICK_GLOBAL_CACHE`\

Author: `<nOob alert>`
Refactored: 
And jap i typoed the env key,

is found.

- `_Resolve()` gets called with the data between the
  [@](https://openusd.org/release/glossary.html#usdglossary-asset) symbols.
- `_Resolve()` checks if the path is an AYON URI path.
  - **Yes?** Then we get the current context (because in this resolver, the
    resolver context interacts with the
    [AyonCppApi](https://github.com/ynput/ayon-cpp-api/) and not the Resolver).
    - We ask the ResolverContext to return the path to us, and the
      ResolverContext calls the `getAsset()` function in the
      `ResolverContextCache`.
    - The `ResolverContextCache` will then first check the PreCache and then the
      Responsible cache. If the `ResolverContextCache` finds the asset, it will
      be returned as a struct. If the `ResolverContextCache` does not find an
      asset, it will call the **AyonCppApi** and request the asset information
      from the server.
  - **No?** Then we check if the AssetIdentifier was already registered in the
    CommonCache, and if so, we will return the cached entry. If not, we will
    resolve the path against the operating file system the same way that USD
    Default Resolver does it.

> ### Asset Identifier / Behavior:

Author: Sorry, but I don't quite understand what you mean. 
I guess you mean we should discuss how identification and communication are being made via a URI? and what the C++ API does.

> ### Asset Identifier / Behavior:

Author: not sure if still valid. 
but also moved over to the other branch 
https://github.com/ynput/ayon-usd-resolver/pull/25#issuecomment-2343288296

The AssetIdentifier or AssetPath is always used by the resolver to convert an
AYON path to a path on disk. The resolver needs some information in the path to

> figure out what asset you want.

Reviewer: picky, correct me if I'm wrong but I'd suggest to rephrase it this way.
The resolver expects a valid  AYON URI, which needs to match that:
* scheme: "ayon"
* path: path to the product folder 
* query: product name, version and representation fixed to usd (with the different options).

![image](https://github.com/user-attachments/assets/49ca8b38-bfb1-4aa6-a6f1-e92cdcc53b70)

Author: I didn't add a picture, but I took your bullet points and added the "old" data to them to expand the query parameter options a bit.

resolver AyonUsdResolver will be used.

but you have a bit more fine grain controle if you want to.

```py
from pxr import Ar
from usdAssetResolver import AyonUsdResolver
Ar.SetPreferredResolver("AyonUsdResolver")
```

Get a resolver via `Usd.Ar` api you will need to get a explicit context to edit
the global context as Ar.GetResolver will return a higher order class and not
AyonUsdResolver class

```py
resolver = Ar.GetResolver()
context = AyonUsdResolver.ResolverContext()
```

If you want fine control over the resolver you will be able to get the connected
context via `GetConnectedContext`\
this is also the recommended way to get access to the dropCache function for the
resolver

```py
explicit_resolver = AyonUsdResolver.Resolver()
explicit_resolver_context = explicit_resolver.GetConnectedContext()
```

`from usdAssetResolver import AyonUsdResolver`python bindings for the
AyonUsdResolver.

`Ar.SetPreferredResolver("AyonUsdResolver")` this command allows you to set a
preferred Resolver. if you use the AyonUsd addon you wont need to set this line
as we register the resolver as the main resolver. but if you have a different
resolver set as Main you might need to use this command.

> `resolver = Ar.GetResolver()` there are multiple ways to get access to an

Reviewer 2: ```suggestion
`resolver = Ar.GetResolver()` there are multiple ways to get access to a
```

resolver instance but using the "Default" Usd option is usually the best idea as
it will return the resolver for the current context.

`context = AyonUsdResolver.ResolverContext()` accessing an ResolverContext
instance is useful if you want to manipulate the Global resolver cache.

### Pinning

the AyonUsdResolver has a feature called Pinning Support.\

> in short pinning support allows you to load a pinning File and disconnect the

Reviewer 2: ```suggestion
In short _pinning support_ allows you to load a pinning file and disconnect the
```

AyonCppApi.\
In the implementation this boils down to a static Memory cache that bypasses the
resolver Logic.\

> why would you do that ? When running a resolver on a farm with many many
> systems(Render Nodes) you might not want them all to call the Ayon server to
> avoid impacting the server performance.

Reviewer 2: ```suggestion
:::note Why use pinning?
When running a resolver on a farm with many workers (render nodes) 
you might not want them all to call the Ayon server to avoid impacting 
the server performance. This is because USD Resolvers will resolve paths
one by one, potentially resulting in many calls.
:::
```

**How to use it ?**\
There are 3 Env variables that control the Pinning file support

`ENABLE_STATIC_GLOBAL_CACHE`\
Enables or Disables static Cache creation. This effectively allows you to enable
or disable Pinning file support.\
as you might want to have a pinning file in your env vars but for e.g debugging

> you don't want to activate the feature.

Reviewer 2: Could you explain more on this? Not sure why or why not?

Author: moved to 
https://github.com/ynput/ayon-usd-resolver/pull/25#issuecomment-2343280880

`PINNING_FILE_PATH`\

> this is a path to the pinning file.

Reviewer 2: ```suggestion
This is a path to the pinning file.
```

Is this always a full path? If so, should it elaborate on that? if it supports relative paths, then "relative" to what? :)

`PROJECT_ROOTS`\

> when running the resolver against the AYON server the CppApi will query the

Reviewer 2: ```suggestion
When running the resolver against the AYON server the CppApi will query the
```

> `Get Project Site Roots` endpoint and get a Dict[str,str] of

Reviewer 2: ```suggestion
`Get Project Site Roots` endpoint and get a `Dict[str,str]` of
```

{root[RootOverwriteKey]}=Val overwrites.\
when running with a pinning file you will need to set this Dict[str,str] as an
ENV variable. e.g:`{Key}:{Path},{Key}:{Path}"` it's not a problem to have
duplicates in the keys but the Cache stores the data as an Unordered_Map so it
will end up deduplicating the Keys. But you can't have spaces in the list as
they are not removed and will be interpreted as part of the Key or Value.

**example**\
setup the resolver for pinning support. we empty all the AYON c++ api keys just
for example you can simply not set them.

```bash
export AYON_API_KEY=""
export AYON_SITE_ID=""
export AYON_SERVER_URL=""
export AYON_PROJECT_NAME=""

export AYONLOGGERLOGLVL=""
export AYONLOGGERFILELOGGING=""
export AYONLOGGERFILEPOS=""

export ENABLE_STATIC_GLOBAL_CACHE="true"
export PINNING_FILE_PATH="/path/to/pinning.json"
export PROJECT_ROOTS="key:/val,key:/val"

export TF_DEBUG=""
```

```py

> stage = Usd.Stage.Open(pinning_file_json["ayon_pinning_data_entry_sceene"])

Reviewer 2: Typo `sceene`?

Also, why do we define the scene file in the JSON?

Author: move to 
https://github.com/ynput/ayon-usd-resolver/pull/25#issuecomment-2343280605

print(stage.ExportToString())
```

PS: its interesting to know that when you generate a pinning file via the
AYON-USD addon the json file will have a key named

> `ayon_pinning_data_entry_sceene`.\

Reviewer 2: Typo.
```suggestion
`ayon_pinning_data_entry_scene`.\
```

This should always be the path used to open the stage Otherwise the pinning file
might not have the correct AssetIdentifier stored.\
aka: if this key points to a local path and you use an URI that points to the

> local path the resolver wont be able to resolve.

Reviewer 2: I don't understand this :)

Author: https://github.com/ynput/ayon-usd-resolver/pull/25#issuecomment-2343280276

### Controlling the cache

`context = AyonUsdResolver.ResolverContext()` there are multiple ways to control
the cache of a resolver. but if your resolver uses the global cache you can
simply create a new ResolverContext and access the cache control functions to
affect the global cache.\

> this does not work if you disconnected the Global cache from your resolver.

Reviewer 2: ```suggestion
This does not work if you disconnected the Global cache from your resolver.
```

> `context.deleteFromCache(AssetIdentifier)` this would allow you to delete an
> individual cached entry.\

Reviewer 2: ```suggestion
`context.deleteFromCache(AssetIdentifier)` delete an individual cached entry.
```

> `context.clearCache()` allows you to clear the connected cache. it is important
> to understand that by default a Resolver will be connected to the global cache
> and this call will delete all entry's in the global cache not just the ones that
> where added by this resolver.\

Reviewer 2: ```suggestion
`context.clearCache()` clear the connected cache. 

:::note
It is important to understand that by default a Resolver will be connected to the global cache
and this call will delete all entries in the global cache not just the ones that
where added by this resolver.
:::
```

`context.dropCache()` allows you do disconnect from the current cache object and
initialize a new one. This can be useful if you want to disconnect from the
global cache and have a Resolver local cache instead.

`explicit_resolver = AyonUsdResolver.Resolver()` if you need to create an
explicit resolver because you want to e.g pass a specific instance into a stage
or you want to disconnect from the Global Cache you should access the Context
connected to this specific resolver.\
`explicit_resolver_context = explicit_resolver.GetConnectedContext()` all the
other functions stay the same.

**Examples**

Deleting cached entry's

```py
resolver = Ar.GetResolver()
context = AyonUsdResolver.ResolverContext()

# Delete From the Global Cache

> AssetIdentifier: str="ayon://{ProjectName}/{path/to/ayon/folder}?product={FileName}&version=latest&representation=usd"
> context.deleteFromCache(AssetIdentifier)

Reviewer 2: ```suggestion
asset_identifier: str="ayon://{ProjectName}/{path/to/ayon/folder}?product={FileName}&version=latest&representation=usd"
context.deleteFromCache(asset_identifier)
```

# Clear the Global Cache
context.clearCache()
```

Disconnecting from the global cache

```py
explicit_resolver = AyonUsdResolver.Resolver()
explicit_resolver_context = explicit_resolver.GetConnectedContext()

# Drooping the cache so that the resolver will generate its own dedicated cache instance
explicit_resolver_context.dropCache()

# Delete from cache and clearCache will also work with an explicit_resolver_context
```

## Resolver Developer Docs

> **Naming BuildPluingins**

Reviewer 2: ```suggestion
**Naming Build Plugins**
```

> BuildPluingins Naming schema is as follows:

Reviewer 2: ```suggestion
Build Plugins Naming schema is as follows:
```

`{AppName}{PlatfromName}/{AppName}{AppVersion}_Py{pythonVersion}_{PlatfromName}.cmake`

> PlatfromName options = {Win, Linux, Mac, specific Os Name}

Reviewer 2: ```suggestion
PlatformName options = {Win, Linux, Mac, specific Os Name}
```

### All About this Project

> **Dependencys**

Reviewer 2: ```suggestion
**Dependencies**
```
